### PR TITLE
OCPEDGE-2408, OCPEDGE-2410: Add AdaptableTopology feature gate and enum to Infrastructure API

### PR DESCRIPTION
**Pull request by jeff-roche**
## Summary

- Add `AdaptableTopology` feature gate (SelfManaged, DevPreviewNoUpgrade) to the Infrastructure API
- Add `Adaptable` enum value to `controlPlaneTopology` and `infrastructureTopology` fields, gated behind `AdaptableTopology`
- Replace static `kubebuilder:validation:Enum` with `FeatureGateAwareEnum` markers so enum validation respects feature gate state
- Add compound `requiredFeatureGate=DualReplica;AdaptableTopology` marker for `controlPlaneTopology` to handle the case where both gates are enabled

## Details

Adaptable Topology enables clusters to dynamically adjust control-plane and infrastructure behavior based on current node count. This PR adds the foundational API types:

1. **Feature gate** — `AdaptableTopology`, scoped to `SelfManaged` clusters at `DevPreviewNoUpgrade`. Not enabled for Hypershift.
2. **Enum constant** — `AdaptableTopologyMode = "Adaptable"` on `TopologyMode`.
3. **Validation markers** — `FeatureGateAwareEnum` annotations on both topology fields that expand the allowed enum values when the gate is enabled. The base (ungated) annotation replaces the previous `kubebuilder:validation:Enum` to avoid overriding feature-gate-aware validation.

Enhancement: https://github.com/openshift/enhancements/pull/1905

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: config/v1/types_infrastructure.go [hunk 1]
```diff
@@ -104,9 +104,13 @@ type InfrastructureStatus struct {
 	// its components are not visible within the cluster.
 	// The 'HighlyAvailableArbiter' mode indicates that the control plane will consist of 2 control-plane nodes
 	// that run conventional services and 1 smaller sized arbiter node that runs a bare minimum of services to maintain quorum.
+	// The 'Adaptable' mode indicates the cluster dynamically adjusts control-plane behavior based on
+	// current node count.
 	// +kubebuilder:default=HighlyAvailable
 	// +openshift:validation:FeatureGateAwareEnum:featureGate="",enum=HighlyAvailable;HighlyAvailableArbiter;SingleReplica;External
 	// +openshift:validation:FeatureGateAwareEnum:featureGate=DualReplica,enum=HighlyAvailable;HighlyAvailableArbiter;SingleReplica;DualReplica;External
+	// +openshift:validation:FeatureGateAwareEnum:featureGate=AdaptableTopology,enum=HighlyAvailable;HighlyAvailableArbiter;SingleReplica;DualReplica;Adaptable;External
+	// +openshift:validation:FeatureGateAwareEnum:featureGate=AdaptableTopology+DualReplica,enum=HighlyAvailable;HighlyAvailableArbiter;SingleReplica;DualReplica;Adaptable;External
```
[the remark this answers is not in the document]
Comment (jeff-roche) [reply config/v1/types_infrastructure.go]: Removing `DualReplica` breaks DevPreview/CustomNoUpgrade CRDs

File: config/v1/tests/infrastructures.config.openshift.io/AdaptableTopology.yaml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            baremetal:
+              loadBalancer:
+                type: OpenShiftManagedDefault
+            type: BareMetal
+      expected: |
+        apiVersion: config.openshift.io/v1
+        kind: Infrastructure
+        spec:
+          platformSpec:
+            type: BareMetal
```
Comment (jaypoulz) [inline config/v1/tests/infrastructures.config.openshift.io/AdaptableTopology.yaml]: Should we add a test for none platform support as well?

File: config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_infrastructures-SelfManagedHA-CustomNoUpgrade.crd.yaml [hunk 3]
```diff
@@ -1134,6 +1134,7 @@ spec:
                 - HighlyAvailableArbiter
                 - SingleReplica
                 - DualReplica
+                - Adaptable
```
Comment (jaypoulz) [inline config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_infrastructures-SelfManagedHA-CustomNoUpgrade.crd.yaml:1136]: I think we need to add an explanation for each of these in the description block as well (above)

File: machineconfiguration/v1/zz_generated.crd-manifests/0000_80_machine-config_01_controllerconfigs-SelfManagedHA-CustomNoUpgrade.crd.yaml [hunk 4]
```diff
@@ -1427,6 +1427,7 @@ spec:
                         - HighlyAvailableArbiter
                         - SingleReplica
                         - DualReplica
+                        - Adaptable
```
Comment (jaypoulz) [inline machineconfiguration/v1/zz_generated.crd-manifests/0000_80_machine-config_01_controllerconfigs-SelfManagedHA-CustomNoUpgrade.crd.yaml:1429]: the MachineConfig will also need descriptions above

File: features.md [hunk 5]
```diff
@@ -6,6 +6,7 @@
 | MachineAPIOperatorDisableMachineHealthCheckController| | | | | | | |  |
 | MultiArchInstallAzure| | | | | | | |  |
 | ShortCertRotation| | | | | | | |  |
+| AdaptableTopology| | | <span style="background-color: #519450">Enabled</span> | <span style="background-color: #519450">Enabled</span> | | | |  |
```
Comment (jaypoulz) [inline features.md]: This says we're available on hypershift when dev-preview-no-upgrade is set. We don't support hypershift for now, since it's mutually exclusive.

**Review by jaypoulz — CHANGES_REQUESTED** [review]
Just a few notes of things I thought looked funny. Going to ping @eggfoobar for a second pair of eyes.

File: config/v1/tests/infrastructures.config.openshift.io/AdaptableTopology.yaml (hunk 2 above)
Comment (jeff-roche) [reply config/v1/tests/infrastructures.config.openshift.io/AdaptableTopology.yaml]: I think so yes

File: features.md (hunk 5 above)
Comment (jeff-roche) [reply features.md]: Good call, definitely didn't meant to add that. Will drop the hypershift column

File: config/v1/tests/infrastructures.config.openshift.io/AdaptableTopology.yaml (hunk 2 above)
Comment (jeff-roche) [reply config/v1/tests/infrastructures.config.openshift.io/AdaptableTopology.yaml]: Swapped to platform none tests, we can add the baremetal tests back in when we get closer to bare metal support

File: config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_infrastructures-SelfManagedHA-CustomNoUpgrade.crd.yaml (hunk 3 above)
Comment (jeff-roche) [reply config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_infrastructures-SelfManagedHA-CustomNoUpgrade.crd.yaml:1136]: If we do that, it ends up bleeding into descriptions that don't apply to Adaptable. Also worth noting Dual Replica is in this list but not in the description I suspect for the same reason

@eggfoobar do you know if there is something I'm missing here?

File: config/v1/types_infrastructure.go [hunk 6]
```diff
@@ -107,6 +107,7 @@ type InfrastructureStatus struct {
 	// +kubebuilder:default=HighlyAvailable
 	// +openshift:validation:FeatureGateAwareEnum:featureGate="",enum=HighlyAvailable;HighlyAvailableArbiter;SingleReplica;External
 	// +openshift:validation:FeatureGateAwareEnum:featureGate=DualReplica,enum=HighlyAvailable;HighlyAvailableArbiter;SingleReplica;DualReplica;External
+	// +openshift:validation:FeatureGateAwareEnum:featureGate=AdaptableTopology,enum=HighlyAvailable;HighlyAvailableArbiter;SingleReplica;DualReplica;Adaptable;External
 	// +optional
 	ControlPlaneTopology TopologyMode `json:"controlPlaneTopology"`
```
[the remark this answers is not in the document]
Comment (eggfoobar) [reply config/v1/types_infrastructure.go:113]: @jeff-roche There is a consequence of how things get generated which means we need to add an extra annotation for situations like this where two FeatureGates exist for the same value. 

This is what's missing, but the suggestion is wrong, you'll want to add this annotation under adaptable you added, this will generate a `DualReplica+AdaptableTopology` DevPreview file, which you'll want to add the tests for both DualReplica and AdaptableTopology.

```
+openshift:validation:FeatureGateAwareEnum:featureGate=AdaptableTopology,enum=HighlyAvailable;HighlyAvailableArbiter;SingleReplica;DualReplica;Adaptable;External
+openshift:validation:FeatureGateAwareEnum:requiredFeatureGate=DualReplica;AdaptableTopology,enum=HighlyAvailable;SingleReplica;DualReplica;Adaptable;External
```

Feel free to peruse [this PR](https://github.com/openshift/api/pull/2283/changes#diff-07b264a49084976b670fb699badaca1795027d6ea732a99226a5388104f6174fR104) where we originally added Arbiter and DualReplica.

Locally you can run `make integration` to validate the unit tests won't fail.

**Review by eggfoobar — COMMENTED** [review]
Looking good @jeff-roche we're just missing some small additions and tests, you'll need to add some extra files to resolve the verify, I think those are manually added if I remember right.

This is the location of the machine config tests, feel free to copy and modify the DualReplica one.
- MachineConfig location: https://github.com/openshift/api/tree/master/machineconfiguration/v1/tests/controllerconfigs.machineconfiguration.openshift.io

Some of the other stuff should have been generated as well, after you add the machine config stuff, try running the generate again `PROTO_OPTIONAL=1 make update`, we're missing some CustomNoUPgrade CRDs changes.

File: features/features.go [hunk 7]
```diff
@@ -115,6 +115,14 @@ var (
 					enable(inDefault(), inOKD(), inTechPreviewNoUpgrade(), inDevPreviewNoUpgrade()).
 					mustRegister()
 
+	FeatureGateAdaptableTopology = newFeatureGate("AdaptableTopology").
+					reportProblemsToJiraComponent("Adaptable Topology").
+					contactPerson("jaypoulz").
+					productScope(ocpSpecific).
+					enhancementPR("https://github.com/openshift/enhancements/pull/1905").
+					enable(inClusterProfile(SelfManaged), inDevPreviewNoUpgrade()).
```
Comment (JoelSpeed) [inline features/features.go:123]: Only us `inClusterProfile` if you think this will eventually be promoted into HyperShift separately. If it's actually `this will never affect hypershift` you don't need to do this
